Macros called as functions link to a misleading / inaccurate error index.

```rust
fn main() {
    println("");
}
```

Produces this error with the helpful text:

```
error[E0423]: expected function, found macro `println`
 --> src/main.rs:2:5
  |
2 |     println("");
  |     ^^^^^^^ did you mean `println!(...)`?
```

However, if you look at the [error index for E0423](https://doc.rust-lang.org/error-index.html#E0423)...

> A struct variant name was used like a function name.

That is **not useful**.  This *used* to point to [E0425](https://doc.rust-lang.org/error-index.html#E0425):

> An unresolved name was used.

Other E0423 instances:

```sh
test/ui/resolve/issue-18252.stderr:1:error[E0423]: expected function, found struct variant `Foo::Variant`
test/ui/resolve/issue-19452.stderr:1:error[E0423]: expected value, found struct variant `Homura::Madoka`
test/ui/resolve/issue-19452.stderr:7:error[E0423]: expected value, found struct variant `issue_19452_aux::Homura::Madoka`
test/ui/resolve/issue-33876.stderr:1:error[E0423]: expected value, found trait `Bar`
test/ui/resolve/issue-39226.stderr:1:error[E0423]: expected value, found struct `Handle`
test/ui/resolve/issue-6702.stderr:1:error[E0423]: expected function, found struct `Monster`
test/ui/resolve/privacy-struct-ctor.stderr:1:error[E0423]: expected value, found struct `Z`
test/ui/resolve/privacy-struct-ctor.stderr:14:error[E0423]: expected value, found struct `S`
test/ui/resolve/privacy-struct-ctor.stderr:26:error[E0423]: expected value, found struct `xcrate::S`
test/ui/resolve/resolve-hint-macro.stderr:1:error[E0423]: expected function, found macro `assert`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:1:error[E0423]: expected value, found module `a`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:9:error[E0423]: expected value, found module `a`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:17:error[E0423]: expected value, found module `a`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:25:error[E0423]: expected value, found module `a::b`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:34:error[E0423]: expected value, found module `a`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:42:error[E0423]: expected value, found module `a::b`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:50:error[E0423]: expected value, found module `a::b`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:59:error[E0423]: expected value, found module `a::b`
test/ui/resolve/suggest-path-instead-of-mod-dot-item.stderr:67:error[E0423]: expected function, found module `a::b`
test/ui/resolve/tuple-struct-alias.stderr:1:error[E0423]: expected function, found self type `Self`
test/ui/resolve/tuple-struct-alias.stderr:13:error[E0423]: expected function, found type alias `A`
```

The error index should describe E0423 as an "expected (value | function) found something else" error. (E0425 is inaccurate since the name does exist, just in a different namespace.)

--- a/src/librustc_resolve/diagnostics.rs
+++ b/src/librustc_resolve/diagnostics.rs
@@ -967,16 +967,18 @@ one.
 "##,
 
 E0423: r##"
-A `struct` variant name was used like a function name.
+An identifier was used like a function name or a value was expected and the
+identifier exists but it belongs to a different namespace.
 
-Erroneous code example:
+For (an erroneous) example, here a `struct` variant name were used as a
+function:
 
 ```compile_fail,E0423
 struct Foo { a: bool };
 
 let f = Foo();
-// error: `Foo` is a struct variant name, but this expression uses
-//        it like a function name
+// error: expected function, found `Foo`
+// `Foo` is a struct name, but this expression uses it like a function name
 ```
 
 Please verify you didn't misspell the name of what you actually wanted to use
@@ -986,6 +988,30 @@ here. Example:
 fn Foo() -> u32 { 0 }
 
 let f = Foo(); // ok!
+```
+
+It is common to forget the trailing `!` on macro invocations, which would also
+yield this error:
+
+```compile_fail,E0423
+println("");
+// error: expected function, found macro `println`
+// did you mean `println!(...)`? (notice the trailing `!`)
+```
+
+Another case where this error is emitted is when a value is expected, but
+something else is found:
+
+```compile_fail,E0423
+pub mod a {
+    pub const I: i32 = 1;
+}
+
+fn h1() -> i32 {
+    a.I
+    //~^ ERROR expected value, found module `a`
+    // did you mean `a::I`?
+}
 ```
 "##,
 

